types.cc

```diff
@@ -2185,6 +2076,125 @@ void tuple_type_impl::validate(bytes_view v, cql_serialization_format sf) const
+        }
+
+        duration_type_impl::counter_type months, days, nanoseconds;
+        std::tie(months, days, nanoseconds) = t.deserialize_counters(v);
+
+        auto check_counter_range = [] (duration_type_impl::counter_type value, auto counter_value_type_instance,
+                                           std::string_view counter_name) {
+            using counter_value_type = decltype(counter_value_type_instance);
+
+            if (static_cast<counter_value_type>(value) != value) {
+                throw marshal_exception(format("The duration {} ({:d}) must be a {:d} bit integer", counter_name, value,
+                        std::numeric_limits<counter_value_type>::digits + 1));
+            }
+        };
+
+        check_counter_range(months, months_counter::value_type(), "months");
+        check_counter_range(days, days_counter::value_type(), "days");
+        check_counter_range(nanoseconds, nanoseconds_counter::value_type(), "nanoseconds");
+
+        if (!(((months <= 0) && (days <= 0) && (nanoseconds <= 0)) ||
+                    ((months >= 0) && (days >= 0) && (nanoseconds >= 0)))) {
+            throw marshal_exception(format("The duration months, days, and nanoseconds must be all of "
+                                           "the same sign ({:d}, {:d}, {:d})",
+                    months, days, nanoseconds));
+        }
+    }
+    template <typename T> void operator()(const floating_type_impl<T>& t) {
+        if (v.size() != 0 && v.size() != sizeof(T)) {
+            throw marshal_exception(format("Expected {:d} bytes for a floating type, got {:d}", sizeof(T), v.size()));
+        }
+    }
+    void operator()(const simple_date_type_impl& t) {
+        if (v.size() != 0 && v.size() != 4) {
+            throw marshal_exception(format("Expected 4 byte long for date ({:d})", v.size()));
+        }
+    }
+    void operator()(const time_type_impl& t) {
+        if (v.size() != 0 && v.size() != 8) {
+            throw marshal_exception(format("Expected 8 byte long for time ({:d})", v.size()));
+        }
+    }
+    void operator()(const uuid_type_impl& t) {
+        if (v.size() != 0 && v.size() != 16) {
+            throw marshal_exception(format("Validation failed for uuid - got {:d} bytes", v.size()));
+        }
+    }
+    void operator()(const inet_addr_type_impl& t) {
+        if (v.size() != 0 && v.size() != sizeof(uint32_t) && v.size() != 16) {
+            throw marshal_exception(format("Validation failed for inet_addr - got {:d} bytes", v.size()));
+        }
+    }
+    void operator()(const map_type_impl& t) { validate_aux(t, v, sf); }
+    void operator()(const set_type_impl& t) { validate_aux(t, v, sf); }
+    void operator()(const list_type_impl& t) { validate_aux(t, v, sf); }
+    void operator()(const tuple_type_impl& t) { validate_aux(t, v, sf); }
+};
+}
+
+void abstract_type::validate(bytes_view v, cql_serialization_format sf) const {
+    visit(*this, validate_visitor{v, sf});
```

Comment:
One thing we can do later, is to have a validate function for each cql_serialization_format (essentially making it a template parameter) and store the pointer to the function in the connection. This eliminates moving the format around.

Reply:
resolving since it is for another series.